agent/pkg/proto/agent.proto

```diff
@@ -0,0 +1,272 @@
+syntax = "proto3";
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/pulumi/pulumi-kubernetes-operator/agent/pkg/proto";
+
+package agent;
+
+service AutomationService {
+
+  /**
+   * WhoAmI returns detailed information about the currently logged-in Pulumi
+   * identity.
+   */
+  rpc WhoAmI(WhoAmIRequest) returns (WhoAmIResult) {}
+
+  /**
+   * Info returns information about the given stack.
+   */
+  rpc SelectStack(SelectStackRequest) returns (SelectStackResult) {}
+
+  /**
+   * Info returns information about the given stack.
+   */
+  rpc Info(InfoRequest) returns (InfoResult) {}
+
+  /**
+   * SetAllConfig sets multiple configuration values for a stack.
+   */
+  rpc SetAllConfig(SetAllConfigRequest) returns (SetAllConfigResult) {}
+
+  /**
+   * Install installs the Pulumi plugins and dependencies.
+   */
+  rpc Install(InstallRequest) returns (InstallResult) {}
+
+  /**
+   * Preview performs a dry-run update to a stack, returning the expected
+   * changes.
+   */
+  rpc Preview(PreviewRequest) returns (stream PreviewStream) {}
+
+  /**
+   * Refresh updates the resources in a stack to match the current state of the
+   * cloud provider.
+   */
+  rpc Refresh(RefreshRequest) returns (stream RefreshStream) {}
+
+  /**
+   * Up creates or updates the resources in a stack, returning the actual
+   * changes.
+   */
+  rpc Up(UpRequest) returns (stream UpStream) {}
+
+  /**
+   * Destroy deletes all resources in a stack.
+   */
+  rpc Destroy(DestroyRequest) returns (stream DestroyStream) {}
+}
+
+message WhoAmIRequest {}
+
+// WhoAmIResult contains detailed information about the currently logged-in
+// Pulumi identity.
+message WhoAmIResult {
+  string user = 1;
+  repeated string organizations = 2;
+  string url = 3;
+}
+
+message SelectStackRequest {
+  string stackName = 1;
+  optional bool create = 2;
+  optional string secrets_provider = 3;
+}
+
+message SelectStackResult { StackSummary summary = 1; }
+
+message InfoRequest {}
+
+message InfoResult { StackSummary summary = 1; }
+
+// StackSummary is a description of a stack and its current status.
+message StackSummary {
+  string name = 1;
+  google.protobuf.Timestamp last_update = 2;
+  bool update_in_progress = 3;
+  optional int32 resource_count = 4;
+  optional string url = 5;
+}
+
+message SetAllConfigRequest {
+  optional bool path =
+      1; // Parse the keys as paths in a map or list rather than raw strings
```

Comment:
Am I right in thinking this basically resolves the structure config problem by letting you mount any config map you want into the pod and use that as your stack config directly? Does that mean SetAllConfig would only get invoked once per operation?

Reply:
This `SetAllConfig` RPC method closely mirrors the underlying auto api, that takes a `path` bool to indicate that the keys in the provided map are "paths" as opposed to literal config keys. What's special here is that the RPC method is able to resolve "valueFrom" refs that draw values from environment variables and local files.

The Stack controller calls `SetAllConfig` today after extracting values from configmaps and secrets.  In the v2 design, the controller doesn't see the values. The configmaps/secrets are mounted onto the workspace pod (as envs and/or files) and then read via these "valueFrom" refs.